File: 290-word-pattern/word-pattern.js
/**
 * @param {string} pattern
 * @param {string} s
 * @return {boolean}
 */
var wordPattern = function(pattern, s) {
    const words = s.split(/\s+/);
    if (words.length != pattern.length) {
        return false
    }

    const letter_map = new Map();
    const word_map = new Map()

    for (let i = 0; i<pattern.length; i++) {
        const w = words[i]
        const l = pattern[i]
        if (word_map.has(w)) {
            if (word_map.get(w) != l) {
                return false
            }}
        word_map.set(w, l)
        if (letter_map.has(l)) {
            if (letter_map.get(l) != w) {
                return false
            }
        }
        letter_map.set(l,w)
        
    }
    return true
};

File: 290-word-pattern/word-pattern.py
class Solution:
    def wordPattern(self, pattern: str, s: str) -> bool:
        letter_map = {}
        word_map = {}
        words = s.split()
        if len(pattern) != len(words):
            return False

        for i in range(len(pattern)):
            l = pattern[i]
            w = words[i]
            if l in letter_map:
                if letter_map[l] != w:
                    return False
            letter_map[l] = w
            
            if w in word_map:
                if word_map[w] != l:
                    return False
            word_map[w] = l

        return True 